code/tests/generic_321_2.cpp

```diff
@@ -0,0 +1,202 @@
+/*
+Reproducing xfstest generic/321 rename_test()
+
+1. Create file foo under mnt_dir_
+2. Create a directory A, and fsync file foo
+3. Move the file foo to directory A
+4. fsync directory A, and fsync file foo
+
+After a crash, mnt_dir_ should contain only the directory A
+and A should contain just the file foo.
+
+https://github.com/kdave/xfstests/blob/master/tests/generic/321
+*/
+
+
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <string>
+#include <iostream>
+#include <dirent.h>
+#include <cstring>
+#include <errno.h>
+
+#include "BaseTestCase.h"
+#include "../user_tools/api/workload.h"
+#include "../user_tools/api/actions.h"
+#define TEST_FILE_FOO "foo"
+#define TEST_DIR_A "test_dir_a"
+
+
+using fs_testing::tests::DataTestResult;
+using fs_testing::user_tools::api::WriteData;
+using fs_testing::user_tools::api::WriteDataMmap;
+using fs_testing::user_tools::api::Checkpoint;
+using std::string;
+
+#define TEST_FILE_PERMS  ((mode_t) (S_IRWXU | S_IRWXG | S_IRWXO))
+
+namespace fs_testing {
+namespace tests {
+
+
+class Generic321_2: public BaseTestCase {
+ public:
+  virtual int setup() override {
+
+	init_paths();
+
+    //Create file foo in mnt_dir_/
+    const int fd_foo = open(foo_path.c_str(), O_RDWR | O_CREAT, TEST_FILE_PERMS);
+    if (fd_foo < 0) {
+      return -1;
+    }
+
+    // Create test directory A.
+	int res = mkdir(dir_path.c_str(), 0777);
+    if (res < 0) {
+      return -1;
+    }
+
+    // Sync the file foo
+    if (fsync(fd_foo) < 0) {
+    	return -1;
+    }
```

Comment:
Have a sync() at the end of setup, to ensure everything is persisted.

Reply:
Done.